### PR TITLE
winset ternaries and embedded wingets

### PR DESCRIPTION
**Pull request by amylizzle**
Implements the `thing.attribute=value?thing.attribute=newvalue:thing.attribute=othervalue` syntax. 
Implements `thing.attribute=[[whatever.attribute as escaped]]`

oh also I made it so you don't need to explicitly implement `TryGetProperty` for every property, they are now backed by the data field annotations.

### CODE REVIEW

File: OpenDreamClient/Interface/DMF/DMFParser.cs [hunk 1]
```diff
@@ -3,6 +3,7 @@
 using OpenDreamClient.Interface.Descriptors;
 using Robust.Shared.Serialization.Manager;
 using Robust.Shared.Serialization.Markdown.Mapping;
+using YamlDotNet.Core.Tokens;
```
Comment (amylizzle) [inline OpenDreamClient/Interface/DMF/DMFParser.cs:6]:
```suggestion
```

File: OpenDreamClient/Interface/DreamInterfaceManager.cs [hunk 2]
```diff
@@ -23,6 +23,7 @@
 using Robust.Shared.Utility;
 using SixLabors.ImageSharp;
 using System.Linq;
+using System.Text.RegularExpressions;
```
Comment (amylizzle) [inline OpenDreamClient/Interface/DreamInterfaceManager.cs:26]:
```suggestion
```

File: OpenDreamClient/Interface/DMF/DMFParser.cs [hunk 3]
```diff
@@ -185,10 +185,11 @@ public sealed class DMFParser(DMFLexer lexer, ISerializationManager serializatio
...
     }
 
-    private bool TryGetAttribute(out string? element, [NotNullWhen(true)] out string? key, [NotNullWhen(true)] out string? token) {
-        element = null;
-        key = null;
-        token = null;
+    private bool TryGetAttribute([NotNullWhen(true)] out DMFWinSet? winSet) {
+        string? element = null;
+        winSet = null;
+        DMFWinSet? condition;
```
Comment (wixoaGit) [inline OpenDreamClient/Interface/DMF/DMFParser.cs:191]: This isn't used

File: OpenDreamClient/Interface/DMF/DMFParser.cs [hunk 4]
```diff
@@ -185,10 +185,11 @@ public sealed class DMFParser(DMFLexer lexer, ISerializationManager serializatio
...
+        DMFWinSet? elseValue;
```
Comment (wixoaGit) [inline OpenDreamClient/Interface/DMF/DMFParser.cs:192]: This can be inlined as `var out elseValue` on line 236

File: OpenDreamClient/Interface/DMF/DMFParser.cs [hunk 5]
```diff
@@ -221,14 +222,27 @@ public sealed class DMFParser(DMFLexer lexer, ISerializationManager serializatio
...
                     Error($"Invalid attribute value ({valueText})");
             } else if (!Check(TokenType.Value))
-                if(Check(TokenType.Semicolon)) //thing.attribute=; means thing.attribute=empty string
+                if(Check(TokenType.Semicolon) || Check(TokenType.EndOfFile)) //thing.attribute=; means thing.attribute=empty string
                     valueText = "";
                 else
                     Error($"Invalid attribute value ({valueText})");
+            else if (Check(TokenType.Ternary)) {
+                List<DMFWinSet> ifValues = new();
+                List<DMFWinSet> elseValues = new();
```
Comment (wixoaGit) [inline OpenDreamClient/Interface/DMF/DMFParser.cs:231]: In BYOND are there allowed to be multiple ifValues or more than 2 elseValues?

File: OpenDreamClient/Interface/DreamInterfaceManager.cs [hunk 6]
```diff
@@ -484,8 +484,15 @@ internal sealed class DreamInterfaceManager : IDreamInterfaceManager {
     }
 
     public void WinSet(string? controlId, string winsetParams) {
-        DMFLexer lexer = new DMFLexer(winsetParams);
-        DMFParser parser = new DMFParser(lexer, _serializationManager);
+        DMFLexer lexer;
+        DMFParser parser;
+        try{
+            lexer = new DMFLexer(winsetParams);
```
Comment (wixoaGit) [inline OpenDreamClient/Interface/DreamInterfaceManager.cs:490]:
```suggestion
        DMFParser parser;
        try{
            var lexer = new DMFLexer(winsetParams);
```

File: OpenDreamClient/Interface/DreamInterfaceManager.cs [hunk 7]
```diff
@@ -498,34 +505,92 @@ internal sealed class DreamInterfaceManager : IDreamInterfaceManager {
             return true;
         }
 
+        string HandleEmbeddedWinget(string? controlId, string value) {
+            string result = value;
+            int startPos = result.IndexOf("[[");
```
Comment (wixoaGit) [inline OpenDreamClient/Interface/DreamInterfaceManager.cs:510]: IndexOf is culture sensitive. Probably not a big deal with `[[` and `]]` but still good practice to specify how to compare the strings.
```suggestion
            int startPos = result.IndexOf("[[", StringComparison.Ordinal);
```

File: OpenDreamClient/Interface/DreamInterfaceManager.cs [hunk 8]
```diff
@@ -498,34 +505,92 @@ internal sealed class DreamInterfaceManager : IDreamInterfaceManager {
...
+            while(startPos > -1){
+                int endPos = result.IndexOf("]]", startPos);
```
Comment (wixoaGit) [inline OpenDreamClient/Interface/DreamInterfaceManager.cs:512]:
```suggestion
                int endPos = result.IndexOf("]]", startPos, StringComparison.Ordinal);
```

File: OpenDreamClient/Interface/DreamInterfaceManager.cs [hunk 9]
```diff
@@ -498,34 +505,92 @@ internal sealed class DreamInterfaceManager : IDreamInterfaceManager {
...
+                string inner = result.Substring(startPos+2, endPos-startPos-2);
+                string[] elementSplit = inner.Split('.');
+                string innerControlId = controlId ?? "";
+                if(elementSplit.Length > 1){
+                    innerControlId = (string.IsNullOrEmpty(innerControlId) ? "" : innerControlId+".")+string.Join(".", elementSplit[..^1]);
+                    inner = elementSplit[^1];
+                }
+                string innerResult = WinGet(innerControlId, inner);
+                result = result.Substring(0, startPos) + innerResult + result.Substring(endPos+2);
+                startPos = result.IndexOf("[[");
```
Comment (wixoaGit) [inline OpenDreamClient/Interface/DreamInterfaceManager.cs:524]:
```suggestion
                startPos = result.IndexOf("[[", StringComparison.Ordinal);
```

File: OpenDreamClient/Interface/DreamInterfaceManager.cs [hunk 10]
```diff
@@ -498,34 +505,92 @@ internal sealed class DreamInterfaceManager : IDreamInterfaceManager {
...
 
             if (CheckParserErrors())
                 return;
 
             // id=abc overrides the elements of other winsets without an element
-            string? elementOverride = winSets.FirstOrNull(winSet => winSet.Element == null && winSet.Attribute == "id")?.Value;
+            string? elementOverride = null;
+            foreach(var winSet in winSets)
+                if(winSet.Attribute == "id" && winSet.Element == null)
+                    elementOverride = HandleEmbeddedWinget(controlId, winSet.Value);
```
Comment (wixoaGit) [inline OpenDreamClient/Interface/DreamInterfaceManager.cs:539]: Why this change? Does having multiple `id=...` do anything?

File: OpenDreamClient/Interface/DreamInterfaceManager.cs [hunk 11]
```diff
@@ -498,34 +505,92 @@ internal sealed class DreamInterfaceManager : IDreamInterfaceManager {
...
-                        element.PopulateElementDescriptor(node, _serializationManager);
+                    if(winSet.IfValues is not null) {
+                        InterfaceElement? conditionalElement = FindElementWithId(elementId);
+                        if(conditionalElement is null)
+                            _sawmill.Error($"Invalid element on ternary condition \"{elementId}\"");
+                        else
+                            if(conditionalElement.TryGetProperty(winSet.Attribute, out var conditionalCheckValue) && conditionalCheckValue.Equals(winSet.Value, StringComparison.InvariantCultureIgnoreCase)) {
+                                foreach(DMFWinSet IfValue in winSet.IfValues) {
+                                    string? IfElementId = IfValue.Element ?? elementId;
+                                    InterfaceElement? IfElement = FindElementWithId(IfElementId);
```
Comment (wixoaGit) [inline OpenDreamClient/Interface/DreamInterfaceManager.cs:559]: These should be renamed to `ifValue`/`ifElementId`/`ifElement`

File: OpenDreamClient/Interface/DreamInterfaceManager.cs [hunk 12]
```diff
@@ -498,34 +505,92 @@ internal sealed class DreamInterfaceManager : IDreamInterfaceManager {
...
+                                        };
+                                        IfElement.PopulateElementDescriptor(node, _serializationManager);
+                                    } else {
+                                        _sawmill.Error($"Invalid element on ternary \"{IfElementId}\"");
+                                    }
+                                }
+                            } else if (winSet.ElseValues is not null){
+                                foreach(DMFWinSet ElseValue in winSet.ElseValues) {
+                                    string? ElseElementId = ElseValue.Element ?? elementId;
+                                    InterfaceElement? ElseElement = FindElementWithId(ElseElementId);
```
Comment (wixoaGit) [inline OpenDreamClient/Interface/DreamInterfaceManager.cs:572]: These should be renamed to `elseValue`/`elseElementId`/`elseElement`

**Review by wixoaGit — CHANGES_REQUESTED** [review]

File: OpenDreamClient/Interface/DMF/DMFParser.cs (hunk 5 above)
Comment (amylizzle) [reply OpenDreamClient/Interface/DMF/DMFParser.cs:231]: Yup, you can do `condition=value?thing1=1;thing2=2:thing1=0;thing2=0`

File: OpenDreamClient/Interface/DreamInterfaceManager.cs (hunk 10 above)
Comment (amylizzle) [reply OpenDreamClient/Interface/DreamInterfaceManager.cs:539]: Id doesn't need to be the first element for it to work, and naturally goon has several instances where it isn't

Comment (wixoaGit) [reply OpenDreamClient/Interface/DreamInterfaceManager.cs:538]: `FirstOrNull()` doesn't return the first value, it returns the earliest value that matches the predicate

File: OpenDreamClient/Interface/DMF/DMFParser.cs (hunk 5 above)
Comment (wixoaGit) [reply OpenDreamClient/Interface/DMF/DMFParser.cs:229]: That's still only 1 condition and 2 branches

Comment (amylizzle) [reply OpenDreamClient/Interface/DMF/DMFParser.cs:229]: Oh I see what you mean, sorry this var is poorly named. The ifValues are the statements evaluated when the condition is true.

File: OpenDreamClient/Interface/DreamInterfaceManager.cs (hunk 10 above)
Comment (amylizzle) [reply OpenDreamClient/Interface/DreamInterfaceManager.cs:538]: Oh there was also an issue with using FirstOrNull() when DMFWinset was no longer a struct? I don't super remember, I'll check it

Comment (amylizzle) [reply OpenDreamClient/Interface/DreamInterfaceManager.cs:538]:
```
The type 'DMFWinSet' must be a non-nullable value type in order to use it as parameter 'TSource' in the generic type or method 'Extensions.FirstOrNull<TSource>(IEnumerable<TSource>, Func<TSource, bool>)'[CS0453](https://msdn.microsoft.com/query/roslyn.query?appId%3Droslyn%26k%3Dk%28CS0453%29)
```
Seems like you can't do `FirstOrNull()` on classes, only structs. I do not know why.

File: OpenDreamClient/Interface/DMF/DMFParser.cs (hunk 5 above)
Comment (amylizzle) [reply OpenDreamClient/Interface/DMF/DMFParser.cs:229]: renamed to TrueStatements and FalseStatements, plus code doc for extra clarity

File: OpenDreamClient/Interface/DreamInterfaceManager.cs (hunk 10 above)
Comment (wixoaGit) [reply OpenDreamClient/Interface/DreamInterfaceManager.cs:538]: `FirstOrNull()` is an RT extension for use with value types. Looks like `FirstOrDefault()` would be what you want here.

**Review by wixoaGit — APPROVED** [review]
